087-external-certificate-manager.md

```diff
@@ -0,0 +1,294 @@
+
+> ![Issuing new component end-entity certificates](./images/087-new-ee-certs.png)
+>
+> Fig 1: Proposed workflow when cert-manager issues new component end-entity certificates
+
+#### Handling CA key replacements
+
+During a CA key replacement the user must pause Strimzi reconciliation until the following are complete:
+* A new CA key has been generated
+* A new CA public certificate has been issued
+* The Secret provided to Strimzi containing the CA public certificate has been updated
+
+Once these are complete Strimzi can be resumed.
+During the next reconciliation Strimzi will:
+1. Copy over the new CA public cert (keeping the old one as described previously) and update the hash and generation annotations.
+2. Roll the Kafka pods to trust the new CA cert, incrementing the `strimzi.io/cluster-ca-key-generation` annotation on the pods.
+3. Copy over the new Kafka certificates, updating the `strimzi.io/cluster-ca-cert-generation` annotations on the Secrets.
+4. Roll the Kafka pods to use the new Kafka certificates, updating the `strimzi.io/cluster-ca-cert-generation` annotations on the Pods.
+5. Then on the next reconciliation, since the Kafka pods now have correct cert and key generation, copy over the new operator certificate.
+6. Since the Kafka pods now have correct cert generation, remove the old CA public cert from the `<CLUSTER_NAME>-cluster-ca-cert` Secret
+
+> ![Renewing the cluster CA public cert](./images/087-cert-renewals.png)
+>
+> Fig 2: Existing and proposed workflow when the user provides a new cluster CA public cert
+
+### User Operator/Clients CA
+
+When a new Kafka cluster is created Strimzi will copy the CA public cert from the Secret identified in `clientsCa.publicCert` to the `<CLUSTER_NAME>-clients-ca-cert` Secret.
+Strimzi will add two annotations to the `<CLUSTER_NAME>-cluster-ca-cert` Secret:
+* `strimzi.io/ca-cert-hash` with the value being a hash of the certificate.
+* `strimzi.io/ca-cert-generation` initially set to 0.
+
+#### Issuing KafkaUser certificates
+
+When a `KafkaUser` with `spec.authentication.type` set to `tls` is creates Strimzi  will create a `Certificate` custom resource.
+Strimzi will specify the required CN/SANS in the `Certificate` resource for the user certificate.
+Strimzi will specify the Secret for the certificate to be stored in as the existing name for the certificate Secret suffixed with `-cm`, for example `<USER_NAME>-cm`.
+Strimzi will request the certificate in both PEM and PKCS12 format.
+Strimzi will wait for the usual operation timeout during the reconciliation loop for the `Certificate` status to indicate that the certificate has been issued before continuing.
+Once the certificate has been issued, Strimzi will copy the certificate across from the cert-manager provided Secret into its own existing Secret.
+Strimzi will also copy across the ClientsCA public certificate.
```

Comment:
Do we still want to do the copying here? Or do we want the user to get it from the CM or (whoever actually produces it) directly? For the server certificates and CAs, we depend on them so we want to copy them out. The user certificate secrets actually exist only for the users. We do not care about them much otherwise. So I wonder if keeping them twice makes sense?

But maybe we still need it to track the generations etc.